### ISSUE TITLE
JupyterLab `4.0.0b1` breaks Notebook 7 `TopBar` items

### ISSUE THREAD

**Issue by jtpio** [jupyterlab/jupyterlab#14366]
## Description

In https://github.com/jupyter/notebook/pull/6838 we noticed that updating from `4.0.0b0` to `4.0.0b1` breaks some of the `TopBar` toolbar items defined via the settings system (`jupyter.lab.toolbars`):

![image](https://user-images.githubusercontent.com/591645/230625673-7a22f5fe-def0-40f9-886a-dbddc7a15fcf.png)

## Reproduce

See discussion in https://github.com/jupyter/notebook/pull/6838#issuecomment-1500342416.

## Context

As mentioned in https://github.com/jupyter/notebook/pull/6838, it seems related to a settings PR that landed in `4.0.0b1`.

Quoting @brichet:

> Actually, the problem comes from another part of https://github.com/jupyterlab/jupyterlab/pull/14195:

> https://github.com/jupyterlab/jupyterlab/blob/6d9217cc9681879dd822e6aaa5cedad2a648f99c/packages/settingregistry/src/settingregistry.ts#L394-L403

> Everything works fine if I avoid this part for `@jupyterlab/application-extension:top-bar` plugin. And I don't know why it works if the `IDocumentManager` is not a dependency.

cc @fcollonval @brichet since you worked on that PR.

**Comment by jtpio** [comment]
Opening a new issue here in JupyterLab so it can get more attention, since it is currently blocking further updates of Notebook 7.

**Comment by brichet** [comment]
Before the use of cache for settings, the settings were fetched again form server. This fetch is delayed by 100 ms using a `Throttler`:
https://github.com/jupyterlab/jupyterlab/blob/25658adabce206725cafe42161181ba43ae87bcd/packages/apputils-extension/src/settingconnector.ts#L36

Using the cache, the delay does not exist anymore, and the widget factory has not been set yet.
If we artificially add a sleep of 100 ms before the load from cache, everything works fine.
We were lucky it used to work without trouble before, it was just a matter of scheduling.

**Comment by brichet** [comment]
A workaround for Notebook while this is not fixed is to force playing again the transform function after the `widgetFactory` is added. The following fixes the issues for the `4.0.0b1` upgrade, but must be fixed in JupyterLab:

```ts
if (toolbarRegistry) {
  toolbarRegistry.addFactory('TopBar', 'checkpoint', (toolbar) => {
    const widget = new Widget({ node });
    widget.id = DOMUtils.createDomID();
    widget.addClass('jp-NotebookCheckpoint');
    return widget;
  });
  if (settingsRegistry) {
    settingsRegistry.load(
      '@jupyterlab/application-extension:top-bar',
      true
    );
  }
}
```
